Make the GMG solver the new default solver.

gassmoeller opened the pull request:
For now this is just a trial to see which tests break.

bangerth commented:
Get out the popcorn!

bangerth commented:
918 of 1025 tests are failures -- not bad, you broke ~90% of all tests :-)

bangerth commented:
And the test output diff is 100 MB!

gassmoeller commented:
I think this is now ready for a first review. I still want to look through the test output one more time, but I would appreciate feedback on the way I implemented the changes and the fallback functionality in the code.

tjhei commented on source/simulator/helper_functions.cc:
```diff
@@ -2117,6 +2118,57 @@ namespace aspect
+
+        pcout << "Warning, the GMG preconditioner is not supported for models locally conservative discretization. Disabling GMG preconditioner." << std::endl;
+      }
+    else if (geometry_model->get_periodic_boundary_pairs().size() > 0)
```
This is no longer true:v
https://github.com/geodynamics/aspect/blob/b2a9312199df918853086989fed6f4d2fb63b199/source/simulator/stokes_matrix_free.cc#L2389C1-L2410C6

gassmoeller commented on source/simulator/parameters.cc:
```diff
@@ -1327,7 +1327,7 @@ namespace aspect
     prm.enter_subsection ("Material model");
     {
-      prm.declare_entry ("Material averaging", "none",
+      prm.declare_entry ("Material averaging", "project to Q1 only viscosity",
```
I have to think about if this is the right thing to do, because it silently changes the averaging even for models that use AMG preconditioner. I also have to check the cookbooks that discuss averaging to make sure they do use the correct averaging and do not contradict this new default value.

tjhei commented on include/aspect/material_model/interface.h:
```diff
@@ -712,7 +712,8 @@ namespace aspect
         geometric_average_only_viscosity,
-        project_to_Q1_only_viscosity
+        project_to_Q1_only_viscosity,
+        solver_default
```
``default_averaging``?

tjhei commented:
Can you check cookbooks/inner_core_convection/?

tjhei commented:
broken:
- cookbooks/morency_doin_2004: linear solver failure
- cookbooks/inner_core_convection: linear solver failure

gassmoeller commented:
I fixed the cookbooks by specifying AMG for now. Everything else is passing? Does that mean we are ready for flipping the big switch? :smile:

tjhei commented:
can you check:
- cookbooks/subduction_initiation/
- cookbooks/mantle_convection_with_continents_in_annulus/
- cookbooks/crustal_deformation/
- cookbooks/allken_et_al_2012_rift_interaction
- benchmarks/buiter_et_al_2008_jgr (takes a long time to finish)

gassmoeller commented:
Ok, I have update the PR as discussed. I have adjusted two solver default settings to be more reasonable for GMG:
- `GMRES solver restart length = 100`: The GMG solver needs less memory, so we can afford to increase the restart length. While it will make individual iterations more expensive, you can see in many tests that it decreases the iteration count by up to a factor of 2 if they need many iterations (e.g. `200+7` -> `55` in `tests/adiabatic_plate_function`). It should also benefit the iteration count for AMG
- `Number of cheap Stokes solver steps = 1000`: The GMG cheap iterations do not have the same limitations as the AMG iterations, and GMG cheap should be the preferred iteration type. Increase the number to 1000 (= the same as our default expensive limit) to make sure we at least try the cheap iterations sufficiently, before switching to expensive.

Regarding the cookbooks:
- `subduction initiation` and `buiter_...` and `mantle_convection_with_continents_in_annulus` work fine with the changes
- `allken_et_al_2012_rift_interaction` and `crustal_deformation` do not converge with GMG. I changed both back to AMG.

I think I am happy with the current state. I would have to squash before merge, but I wanted to keep the test updates separate from each other for review. Opinions?

tjhei commented:
I think this is good to go now, thank you Rene!

gassmoeller commented:
:tada: What a relief.
